mongo.go

```diff
@@ -0,0 +1,472 @@
+package mongods
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/ipfs/go-datastore"
+	"github.com/ipfs/go-datastore/query"
+	dsq "github.com/ipfs/go-datastore/query"
+	logging "github.com/ipfs/go-log/v2"
+	"github.com/jbenet/goprocess"
+	dsextensions "github.com/textileio/go-datastore-extensions"
+	"go.mongodb.org/mongo-driver/bson"
+	"go.mongodb.org/mongo-driver/bson/primitive"
+	"go.mongodb.org/mongo-driver/mongo"
+	"go.mongodb.org/mongo-driver/mongo/options"
+)
+
+var (
+	ErrClosed = errors.New("datastore was closed")
+
+	log = logging.Logger("mongods")
+)
+
+type MongoDS struct {
+	m          *mongo.Client
+	db         *mongo.Database
+	col        *mongo.Collection
+	opTimeout  time.Duration
+	txnTimeout time.Duration
+
+	lock   sync.RWMutex
+	closed bool
+}
+
+var _ datastore.Datastore = (*MongoDS)(nil)
+var _ datastore.Batching = (*MongoDS)(nil)
+var _ datastore.TxnDatastore = (*MongoDS)(nil)
+var _ dsextensions.ExtendedDatastore = (*MongoDS)(nil)
+
+type keyValue struct {
+	Key   string `bson:"_id"`
+	Value []byte `bson:"v"`
+}
+
+func New(ctx context.Context, uri string, dbName string, collName string, opts ...Option) (*MongoDS, error) {
+	m, err := mongo.Connect(ctx, options.Client().ApplyURI(uri))
+	if err != nil {
+		return nil, fmt.Errorf("connecting to MongoDB: %s", err)
+	}
+	db := m.Database(dbName)
+
+	_ = db.CreateCollection(ctx, collName)
```

Comment:
This might seem odd, since on the first write the collection would _usually_ be created.
There's a border case when you make the first write in a MongoDB transaction; if the collection doesn't exist you get: `multiple write errors: [{write errors: [{Cannot create namespace blocks.blocky in multi-document transaction.}]}`

So, that's a really border-border-case but doing this would prevent any kind of problems.